node/src/components/storage.rs

```diff
@@ -143,32 +151,113 @@ impl<B: Value> Default for DeployMetadata<B> {
-            let deploy_hashes = SmallVec::from(block.deploy_hashes().clone());
-            let block_hash =
-                ProtoBlockHash::from_parts(&deploy_hashes, block.header().random_bit());
-            let deploys = task::spawn_blocking(move || deploy_store.get(deploy_hashes))
+        latest_block_height: u64,
+        chainspec_version: Version,
+        current_instant: Timestamp,
+    ) -> BlockProposerState {
+        //
+        // - load BlockProposerState from storage
+        let block_proposer_state_store = self.block_proposer_state_store();
+        let block_proposer_state = task::spawn_blocking(move || block_proposer_state_store.get())
+            .await
+            .expect("should run");
+
+        // If there is not yet a BlockProposerState stored, we should continue with a default.
+        let mut block_proposer_state = match block_proposer_state {
+            Ok(Some(state)) => state,
+            _ => BlockProposerState::default(),
+        };
+
+        let max_ttl = {
+            let chainspec_store = self.chainspec_store();
+            let chainspec = task::spawn_blocking(move || chainspec_store.get(chainspec_version))
                 .await
-                .expect("should run")
-                .into_iter()
-                .map(|result| {
-                    result.unwrap_or_else(|error| panic!("failed to get deploy: {}", error))
-                })
-                .flatten()
+                .expect("should run blocking");
+            let chainspec = match chainspec {
+                Ok(Some(chainspec)) => chainspec,
+                // If we can't get our hands on a chainspec, then we can't get a max_ttl to compare
+                // blocks and deploys against.
+                _ => return block_proposer_state,
+            };
+            chainspec.genesis.deploy_config.max_ttl
+        };
+
+        // deploys, organized by ProtoBlockHash, which have been finalized
+        let mut finalized = HashMap::new();
+
+        //
+        // - get block by height of latest block
+        'iterate_ancestry: for height in (0..=latest_block_height).rev() {
+            let block = {
+                let block_store = self.block_store();
+                let block_by_height_store = self.block_height_store();
+
+                let ancestor_hash = task::spawn_blocking(move || block_by_height_store.get(height))
+                    .await
+                    .expect("should spawn_blocking");
+                let ancestor_hash = match ancestor_hash {
+                    Ok(Some(hash)) => hash,
+                    _ => break 'iterate_ancestry,
+                };
+
+                let block = task::spawn_blocking(move || block_store.get(smallvec![ancestor_hash]))
+                    .await
+                    .expect("should spawn_blocking")
+                    .pop();
+                match block {
+                    Some(Ok(Some(block))) => block,
+                    _ => break 'iterate_ancestry,
```

Comment:
I'm confused: What does `Some(Ok(None))`, `Some(Err(_))` and `None` mean here? Should we log errors in some of these cases?

Reply:
Thanks, I've tried to clarify this.

Reply:
So `None` means the block doesn't exist… then what does `Some(Ok(None))` mean?

Reply:
This layering of `Option<Result<Option<_>>>` is due to the store + vec operations we're doing, I'll rework this to be clearer as to what's going on. Currently `Ok(None)` _actually_ means that the block doesn't exist.